File: 173-binary-search-tree-iterator/173-binary-search-tree-iterator.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class BSTIterator {
    Stack<TreeNode> stk = new Stack<>();
    public BSTIterator(TreeNode root) {
        while(root!=null){
            stk.push(root);
            root=root.left;
        }
    }
    
    public int next() {
        TreeNode node = stk.pop();
        TreeNode temp=node.right;
        while(temp!=null){
            stk.push(temp);
            temp=temp.left;
        }
        return node.val;
    }
    
    public boolean hasNext() {
        return stk.size()!=0;
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.next();
 * boolean param_2 = obj.hasNext();
 */

File: 173-binary-search-tree-iterator/173-binary-search-tree-iterator.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator:

    def __init__(self, root: Optional[TreeNode]):
        def dfs(root):
            if not root: return
            dfs(root.left)
            self.ino.append(root.val)
            dfs(root.right)
        
        self.ino=[]
        self.ptr=0
        dfs(root)
        

    def next(self) -> int:
        ans = self.ino[self.ptr]
        self.ptr+=1
        return ans

    def hasNext(self) -> bool:
        return self.ptr<len(self.ino)


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()